keps/sig-architecture/3138-increase-reliability-bar/README.md

```diff
@@ -0,0 +1,415 @@
+clusters before.
+
+To make Kubernetes (or any other project) sustainable, we should ensure that
+existing architecture, individual features, tests etc. really works as expected.
+If there is work needed to improve reliability (or stability) of those, it
+should come before new features.
+This is often not the case currently. We significantly rely on the heroic
+efforts of very few individuals that are keeping us on the surface with new bugs,
+issues, test flakes etc. This approach isn’t sustainable and doesn’t scale.
+We need to make it a business of everyone in the whole community.
+
+[Production Readines Survey]: https://datastudio.google.com/reporting/2e9c7439-202b-48a9-8c57-4459e0d69c8d/page/GC5HB
+[How to Break your Cluster with networking]: https://www.youtube.com/watch?v=7qUDyQQ52Uk)
+[1]: https://www.youtube.com/watch?v=QKI-JRs2RIE
+[2]: https://www.youtube.com/watch?v=6sDTB4eV4F8
+[this example]: https://github.com/kubernetes/kubernetes/pull/98305
+
+
+### Goals
+
+- Define the high-level areas of investments
+- Define the sustainable process to ensure reliability investments
+- Define the details of the first milestone as a pilot (reducing test flakiness)
+
+
+### Non-Goals
+
+- Providing exact list of long-standing reliability issues (milestone 3) or new
+  reliability investments (milestone 4) - those will be handled by
+  Reliability WG separately
+
+
+## Proposal
+
+This document outlines steps to improve the quality and reliability of Kubernetes.
+Given the maturity level the project is reaching and the non-negligible amount of
+technical debt accumulated over the years, the most important aspect of the
+proposal is to build the culture which prioritizes certain types of contributions
+over purely feature work.
+
+We have to acknowledge that fixing, and especially debugging, reliability or
+stability issues is the kind of work that very few people like doing - for the
+majority it’s something that they would never do if they wouldn’t have to.
+This is especially true when it comes to debugging flaky tests.
+Moreover, while in companies there are multiple ways of enforcing that certain
+types of work is actually happening (e.g. oncall rotations, dedicated sprints, etc.),
+it’s much harder to achieve in the projects driven by the open-source community.
+
+As a result, we’re proposing introducing a process that will allow us to ensure
+that this kind of work that generally attracts fewer developers  will be really
+happening. This means, that for every piece of such work, we will introduce a two
+phase approach:
+1. announce the reliability-related issue to the SIG asking for prioritizing it
+   with a realistic (and non-conservative) deadline.
+   *We really hope that this will be effective in the majority of cases and
+   we will never be forced to use the point (2) special powers.*
+1. if the issue (reported above) is not addressed before the deadline, block
```

Comment:
Proposing to punish SIGs is harmful to the spirit of supportive collaboration that is central to Kubernetes's success, and runs counter to the project's values. The reliability bar can be kept high and raised without compulsion. No potential reliability improvements are worth the loss of community goodwill caused by establishing a punitive system of enforcement.

Reply:
I think I completely don't follow this comment.
This proposal isn't about punishing anyone - I don't know why you're interpreting it as punishment. It's about saying that from the project perspective some things are important than some other things.

If you're working on a project, and there are thing that have to be done but noone is willing to pick them, the end state isn't saying "forget it - let's work on whatever you want to work". That's not how it works. If some works needs to be done, it needs to be done. And we can't replace that work with other kind of work.

You can be arguing that "X isn't important and shouldn't belong to this category" - that's a discussion I'm definitely open to have. But that's not what you're saying.

> No potential reliability improvements are worth the loss of community goodwill caused by establishing a punitive system of enforcement.

I don't see how goodwill is related to it. How is goodwill related to the fact that we're saying that some things are important?
Regarding being "worth of". The whole Kubernetes project (as well as any other project) makes sense, as long as someone wants to use it. If it becomes unreliable, no real users are going to use it, because it won't satisfy their needs. Something else will be created.

The fact that Kubernetes has such traction as it has now, doesn't mean it will have the same traction in 2 years.
If we won't provide feature X, it may prevent some group of users who want it from using it. That's obviously bad. But if the system will become more and more unreliable, we can lose every single user of it.

And again - it's not a punishment. It's just saying that some things are more important than others.

In reality - I would prefer that we never have to use this right. And we can always use exceptions to not use it. But we need to be able to achieve certain things.

I can give you an actual example - if you look at SIG scability charter, we have a right to rollback any PR that is breaking scalability:
https://github.com/kubernetes/community/blob/master/sig-scalability/charter.md#what-can-we-dorequire-from-other-sigs
This was approved by steering many years ago.
Do we want to use it? Of course no. Do regressions happen? They do. Do we immediately rollback - of course no. We ask people to fix things. Did we use that in the last 5 releases (I can probably even say 10 or even more). No (I remember we used it once around ~1.10 or so). Does it hurt the community - I doubt.


Now - I'm definitely happy to start this proposal in "dry-run" mode (without any enforcement) and run it like that for say 1-2 releases and see what happens. If it will work as expected, we can leave it like that. But if it won't work the we should do something.

Reply:
The sig-scalability rollback was never approved by Steering -- you added it back in the time when steering didn't exist and SIG charters did not need approval.  And you've never used it, and if you actually tried you might find that you don't actually have that power.

Reply:
> The sig-scalability rollback was never approved by Steering --

yes - it was approved by steering.

> And you've never used it

Yes I used that power. I blocked myself making CoreDNS a default one back then.

Reply:
Just to back it up:

https://github.com/kubernetes/community/pull/2149 is the PR (there was also a follow up to it linked from there) where you clearly see steering committee approval on the charter.

And this it the revert where we used that power:
https://github.com/kubernetes/kubernetes/pull/68629
FWIW - it's the only case that I'm aware of where we used that (so almost 12 releases ago).

Reply:
to try to refocus the conversation and don't diverge on specifics, Kubernetes, as other big OpenSource projects are a magnet for companies and/or people to add features and enhance their portfolios or CVs... but the reality is that, on the day a day, there are only a few persons doing the dirty job, fix flakes, upgrade dependencies, .... something that doesn't shine and you can brag on tweet or add to your CV and doesn't attract companies, that are not willing to spend time of their employees on those duties.

I think that this is about accountability and responsibility,  a SIG is not only there ot add features and to decide what goes it or not, it has to maintain its area of Interest, add tests, check that work, improve reliability, performance ... I can't see how that is not a fair request, maybe we can reword it, but this is a community project, these things are not handover to the QE department or something like that

Reply:
Two notes: 
1) "we can always use exceptions to not use it", these exceptions should be memorialized in policy (in this enhancement)
2) " I'm definitely happy to start this proposal in "dry-run" mode (without any enforcement) and run it like that for say 1-2 releases and see what happens" I think the rollout process ie dry-run needs to be memorialized here as well.

These are non-trivial changes will high impact across the community, so let's try to incorporate all of this into the plan so that it is clear what people can expect at each stage of the rollout.